### ISSUE TITLE
wrong alphabet exception

### ISSUE THREAD

**Issue by rdikow** [Jstacs/Jstacs#14]
Hi, thanks for GeMoMa - it works really well! I've recently run into an issue where I get the below error. I've been annotating 6 bird genomes and 5 out of the 6 finish without an error, but the 6th returns the below even when I use the same references as all the others. Thanks for any insights!
-Rebecca

Example command (v 1.7.1): 
`GeMoMa -Xmx600G GeMoMaPipeline threads=$NSLOTS outdir=annotation_manacus GeMoMa.Score=ReAlign AnnotationFinalizer.r=NO o=true t=manacus.fasta s=own i=pipra a=pipra.gff g=pipra.fasta`

Error:
```
Exception in thread "main" java.lang.RuntimeException: Did not finish as intended. de.jstacs.data.WrongAlphabetException: Symbol "J" from input not defined in alphabet: [A, R, N, D, C, Q, E, G, H, I, L, K, M, F, P, S, T, W, Y, V, B, Z, X, *]
	at de.jstacs.data.alphabets.DiscreteAlphabet.getCode(DiscreteAlphabet.java:288)
	at de.jstacs.data.AlphabetContainer.getCode(AlphabetContainer.java:643)
	at de.jstacs.data.sequences.ByteSequence.<init>(ByteSequence.java:159)
	at de.jstacs.data.sequences.ByteSequence.<init>(ByteSequence.java:127)
	at de.jstacs.data.sequences.Sequence.create(Sequence.java:643)
	at de.jstacs.data.sequences.Sequence.create(Sequence.java:611)
	at de.jstacs.data.sequences.Sequence.create(Sequence.java:584)
	at projects.gemoma.GeMoMa.addHit(GeMoMa.java:1173)
	at projects.gemoma.GeMoMa.run(GeMoMa.java:992)
	at projects.gemoma.GeMoMaPipeline$JGeMoMa.doJob(GeMoMaPipeline.java:1922)
	at projects.gemoma.GeMoMaPipeline$FlaggedRunnable.run(GeMoMaPipeline.java:1321)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
GeMoMa for species 0 (pipra) split=44 throws an Exception
```

**Comment by JensKeilwagen** [comment]
Hi rebecca,

thanks a lot for your email. I haven't seen this before.

The error clearly states that the amino acid "J" is unknown in our amino acid alphabet. 
"J" encodes for leucine or isoleucine (https://www.ddbj.nig.ac.jp/ddbj/code-e.html).
It seems that "J" is used in a result of the search algorithm that you use  (cf. "projects.gemoma.GeMoMa.addHit(GeMoMa.java:1173)"), which seems to be mmseqs. Could you please write which version you use?

Typically, broken GeMoMaPipeline runs leave some information in the directory "GeMoMa_temp". Here you could check the results of the search algorithm identifying the gene model that has matches including the amino acid "J". This could help to create a minimal example.

GeMoMa uses the amino acids given in the substitution matrix to define the amino acid alphabet. The substitution matrix that we currently use do not have entries for "J" (https://github.com/Jstacs/Jstacs/blob/master/projects/gemoma/test_data/BLOSUM62.txt). This matrix was downloaded from NCBI some time ago to be compatible to blast (ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/BLOSUM62).

Using a substitution matrix that has entries for "J" might solve the problem. I just had a look and found: https://www.ncbi.nlm.nih.gov/IEB/ToolBox/C_DOC/lxr/source/data/BLOSUM62
Based on this, we might be able to create such a substitution matrix in the correct format.

However, I compared our BLOSUM matrix with the one above and found that they differ not only in the column for "J" but also for some individual values. Hence, it would be good to know which substitution matrix was used by the search algorithm.

Best regards, Jens

**Comment by rdikow** [comment]
Thanks so much for the reply! I am trying to dig through the temp files. I uploaded them here (along with a log file) in case you have time to have a look: https://www.dropbox.com/s/w9a8mf21vvurbdz/gemoma-test.tar.gz?dl=0

Thanks again!

**Comment by JensKeilwagen** [comment]
Hi Rebecca,

thanks a lot for the huge amount of data.

I checked the log file and found that you are using mmseqs version 13.45111.
Checking mmseqs substitution matrices: https://github.com/soedinglab/MMseqs2/tree/master/data
I did not find a column for amino acid "J" in BLOSUM62 and some others that I checked.
I also did not find that they have changed the default substitution matrix or anything regarding leucine/isoleucine: https://github.com/soedinglab/MMseqs2/releases

In addition, I had a look into GeMoMaPipeline-16045397510071019339/0/mmseqs.tabular. Only 3 out of more then 12M lines contained a "J" in the alignment:

3573076	gene_16314_1	NW_021940722.1	44.400	27	15	0	15	41	999	1079	2.172E+01	29	-	62	12	-	-	-			PELPSGAGPFDAATTPARLSDATAFPP	PRVPSGPSPRSVSLJPDRRCQATALPP	61	4478
3850325	gene_16314_3	NW_021940722.1	44.400	27	15	0	14	40	999	1079	2.078E+01	29	-	62	12	-	-	-			PELPSGAGPFDAATTPARLSDATAFPP	PRVPSGPSPRSVSLJPDRRCQATALPP	60	4478
8552716	gene_16474_0	NW_021939122.1	35.800	67	39	3	11	75	11486396	11486202	8.653E+00	32	-	70	24	-	-	-			LLARLGLQALLALLPALARGLGLHPRPPCPGPGSAAPALPAGRAAPLWGPAR--GDEYERRYSRAFP	MLRAVGLNAGRG-APPJSRH-GTLPFPGAPGPARAAPECPSPRSGTERGPISPVGEQPESRENKYGP	121	17883582

(first column is the line number starting with 0)

A very simple work around is to remove the corresponding reference transcript to avoid "J" in the search results of mmseqs. However, I would recommend to contact the mmseqs developers. Maybe they can give better advices.

Looking forward to any news.

best regards, Jens

**Comment by JensKeilwagen** [comment]
I've implemented a new paremeter "replace unknown" that allows to replace unknown amino acids in the search results by "X".

https://github.com/Jstacs/Jstacs/commit/c93cc06231b061f4de79509d03174079b202c8a3

Will be available with the next release.

**Comment by rdikow** [comment]
Thank you so much!

**Comment by JensKeilwagen** [comment]
Laura Baxter also reported this problem to me. However, she found another way to avoid this problem:

"I managed to track the source of the issue to nucleotide ambiguity codes in the target genome sequence FASTA file. I replaced all non-ATGC characters in the sequence with N with a perl one-liner, and then everything worked great. Worth remembering in case anyone else has the same issue in the meantime.

The one-liner to replace any non-ATGC character in the sequence with N was:
perl -pe 's/[^AGTC\n]/N/gi unless m/>/;' oldfile.fas > newfile.fas
"

Thanks Laura.
